### PR TITLE
HTTP runtime metrics

### PR DESCRIPTION
**Pull request by bengl**
### What does this PR do?

Add support for runtime metrics over HTTP.

### Motivation

Node.js does not natively support DGRAM Unix domain sockets, so this simplifies configuration and support for cases where the library is intended to communicate with the agent over Unix domain sockets. It also means tracing and runtime metrics can be configured using a single endpoint.

### CODE REVIEW

File: packages/dd-trace/src/metrics.js [hunk 1]
```diff
@@ -49,6 +49,7 @@ module.exports = {
     }
 
     client = new Client({
+      tracingUrl: config.url || new URL(`http://${config.hostname || 'localhost'}:${config.port}`),
```
Comment (rochdev) [inline packages/dd-trace/src/metrics.js:52]: This is unrelated to tracing other than sharing the endpoint.

File: packages/dd-trace/test/dogstatsd.spec.js [hunk 2]
```diff
@@ -1,14 +1,22 @@
 'use strict'
 
-describe('dogstatsd', () => {
+const http = require('http')
+
+describe.only('dogstatsd', () => {
```
Comment (rochdev) [inline packages/dd-trace/test/dogstatsd.spec.js:5]: 😬

File: packages/dd-trace/src/dogstatsd.js [hunk 3]
```diff
@@ -11,6 +13,16 @@ class Client {
   constructor (options) {
     options = options || {}
 
+    if (options.tracingUrl) {
+      this._httpOptions = url.parse(options.tracingUrl)
+      this._httpOptions.agent = null
+      if (this._httpOptions.protocol === 'unix:') {
```
Comment (rochdev) [inline packages/dd-trace/src/dogstatsd.js:19]: Why is this needed? It should be handled by the underlying HTTP client by moving the code from the writer.

File: packages/dd-trace/src/dogstatsd.js [hunk 4]
```diff
@@ -31,14 +43,26 @@ class Client {
...
     if (this._queue.length === 0) return
 
     this._queue = []
 
-    if (this._family !== 0) {
+    if (this._httpOptions) {
+      requestQueue(queue, this._httpOptions, err => {
+        if (err) {
+          log.debug('HTTP error from agent: ' + err.stack)
+          this._httpOptions = null
```
Comment (rochdev) [inline packages/dd-trace/src/dogstatsd.js:57]: This should actually only happen based on the status code. Socket errors mean that the agent couldn't be reached, in which case it might be down or completely unreachable. Only a 404 would confirm that the agent does not support HTTP for runtime metrics.

File: packages/dd-trace/src/dogstatsd.js [hunk 5]
```diff
@@ -104,3 +128,23 @@ class Client {
 }
 
 module.exports = Client
+
+function requestQueue (queue, options, callback) {
```
Comment (rochdev) [inline packages/dd-trace/src/dogstatsd.js:132]: Let's keep the export as the last line so that it's easy to find.

File: packages/dd-trace/src/dogstatsd.js [hunk 6]
```diff
@@ -31,14 +43,26 @@ class Client {
...
     this._queue = []
 
-    if (this._family !== 0) {
+    if (this._httpOptions) {
+      requestQueue(queue, this._httpOptions, err => {
+        if (err) {
+          log.debug('HTTP error from agent: ' + err.stack)
+          this._httpOptions = null
+          this._queue = queue
+          this.flush()
```
Comment (rochdev) [inline packages/dd-trace/src/dogstatsd.js:59]: This is flushing only happening on error?

File: packages/dd-trace/src/dogstatsd.js (hunk 5 above)
Comment (rochdev) [reply packages/dd-trace/src/dogstatsd.js:132]: Also, why is this function needed?

**Review by rochdev — CHANGES_REQUESTED** [review]

File: packages/dd-trace/src/metrics.js (hunk 1 above)
Comment (bengl) [reply packages/dd-trace/src/metrics.js:52]: Yeah I was hoping for a better name here. Any suggestions? Maybe `agentUrl` or `dogstatsdProxyUrl`?

File: packages/dd-trace/test/dogstatsd.spec.js (hunk 2 above)
Comment (bengl) [reply packages/dd-trace/test/dogstatsd.spec.js:5]: Whoops! Will fix.

File: packages/dd-trace/src/dogstatsd.js (hunk 3 above)
Comment (bengl) [reply packages/dd-trace/src/dogstatsd.js:19]: We do _some_ normalization today in `request`, but it takes in an options object, rather than a URL. I _guess_ we could put this into that normalization somehow since it's also happening in writer, but that would mean an extra property on the options object I think. Maybe `url`. I'll give that a try.

File: packages/dd-trace/src/dogstatsd.js (hunk 4 above)
Comment (bengl) [reply packages/dd-trace/src/dogstatsd.js:57]: What should be the behaviour when there's a socket error?

File: packages/dd-trace/src/dogstatsd.js (hunk 5 above)
Comment (bengl) [reply packages/dd-trace/src/dogstatsd.js:132]: If we have multiple queued payloads, we'll send them all sequentially, like we do with UDP. This is to make that easy for HTTP.

File: packages/dd-trace/src/dogstatsd.js (hunk 6 above)
Comment (bengl) [reply packages/dd-trace/src/dogstatsd.js:59]: The _recursive_ flush only happens on error. By setting the two properties above, it forces it to do UDP on the next attempt.

File: packages/dd-trace/src/dogstatsd.js (hunk 5 above)
Comment (rochdev) [reply packages/dd-trace/src/dogstatsd.js:135]: Still not sure I understand, why is it not enough to simply send the request? The queueing happens externally.

File: packages/dd-trace/src/dogstatsd.js [hunk 7]
```diff
@@ -31,14 +41,31 @@ class Client {
...
+            // Inside this if-block, we have connectivity to the agent, but
+            // we're not getting a 200 from the proxy endpoint. Fall back to
+            // UDP and try again.
+            this._httpOptions = null
+            this._queue = queue
+            this.flush()
+          }
+        }
+        if (cb) {
+          cb(err)
```
Comment (rochdev) [inline packages/dd-trace/src/dogstatsd.js:65]: Should this also be logged?

**Review by rochdev — CHANGES_REQUESTED** [review]

File: packages/dd-trace/src/dogstatsd.js (hunk 7 above)
Comment (bengl) [reply packages/dd-trace/src/dogstatsd.js:65]: Yeah, and we're doing that a few lines up. I'll change the log level there to error.

File: packages/dd-trace/src/dogstatsd.js [hunk 8]
```diff
@@ -48,16 +56,37 @@ class Client {
...
     log.debug(`Sending to DogStatsD: ${buffer}`)
 
     socket.send(buffer, 0, buffer.length, this._port, address)
   }
 
-  _sendAll (queue, address, family) {
-    queue.forEach((buffer) => this._send(address, family, buffer))
+  _sendAll (queue, address, family, options) {
+    queue.forEach((buffer) => {
+      if (options) {
```
Comment (rochdev) [inline packages/dd-trace/src/dogstatsd.js:84]: This condition is repeated here and in the parent. Why not just call the right method directly?

Comment (bengl) [reply packages/dd-trace/src/dogstatsd.js:84]: Do you mean like:

```js
if (options) {
  queue.forEach(buffer => this._sendHttp(buffer))
} else {
    queue.forEach(buffer => this._sendUdp(buffer))
}
```

Or do you mean having a `_sendAllHttp` and `_sendAllUdp`?

Comment (rochdev) [reply packages/dd-trace/src/dogstatsd.js:84]: I mean calling the right one in the parent yes. Although looking at it more it does multiple HTTP requests but it should make only 1 since HTTP is much more inefficient compared to UDP.

**Review by rochdev — CHANGES_REQUESTED** [review]

**Review by Qard — APPROVED** [review]

File: packages/dd-trace/src/dogstatsd.js [hunk 9]
```diff
@@ -48,16 +56,38 @@ class Client {
...
 
     log.debug(`Sending to DogStatsD: ${buffer}`)
 
     socket.send(buffer, 0, buffer.length, this._port, address)
   }
 
-  _sendAll (queue, address, family) {
-    queue.forEach((buffer) => this._send(address, family, buffer))
+  _sendAll (queue, address, family, options) {
+    if (options) {
```
Comment (rochdev) [inline packages/dd-trace/src/dogstatsd.js:84]: This is unnecessarily confusing, this condition would be better suited in the existing condition in flush so that it only happens once instead of multiple times. There are no scenarios that would result in an HTTP call except when there are HTTP options.

File: packages/dd-trace/src/exporters/common/writer.js [hunk 10]
```diff
@@ -5,7 +5,7 @@ const log = require('../../log')
 
 class Writer {
   constructor ({ url }) {
-    this._url = url
+    this._url = url.toString()
```
Comment (rochdev) [inline packages/dd-trace/src/exporters/common/writer.js:8]: Why is this change needed?

File: packages/dd-trace/src/metrics.js [hunk 11]
```diff
@@ -48,11 +48,19 @@ module.exports = {
...
 
-    client = new Client({
+    const clientConfig = {
       host: config.dogstatsd.hostname,
       port: config.dogstatsd.port,
       tags
-    })
+    }
+
+    if (config.url) {
```
Comment (rochdev) [inline packages/dd-trace/src/metrics.js:57]: Why not just pass through everything if it's handled in `request`?

File: packages/dd-trace/test/dogstatsd.spec.js [hunk 12]
```diff
@@ -179,4 +223,65 @@ describe('dogstatsd', () => {
     expect(udp6.send.firstCall.args[3]).to.equal(7777)
     expect(udp6.send.firstCall.args[4]).to.equal('::1')
   })
+
+  const udsIt = os.platform() === 'win32' ? it.skip : it
+  udsIt('should support HTTP via unix domain socket', (done) => {
```
Comment (rochdev) [inline packages/dd-trace/test/dogstatsd.spec.js:228]: Let's wrap the test in a condition instead. This makes it easier to manually skip/only when needed. Skipping tests is also usually meant for a test that will be implemented in the future or is temporarily disabled, not a test that will never run which just pollutes the output.

**Review by rochdev — CHANGES_REQUESTED** [review]

File: packages/dd-trace/test/dogstatsd.spec.js (hunk 12 above)
Comment (Qard) [reply packages/dd-trace/test/dogstatsd.spec.js:228]: That has not been my experience. Other test frameworks have a `todo(...)` function for that and `skip(...)` is used specifically for conditional inclusion of tests like platform support.

Comment (rochdev) [reply packages/dd-trace/test/dogstatsd.spec.js:228]: Other test frameworks might use these words for other things, but from the Mocha documentation: "Anything skipped will be marked as [pending](https://mochajs.org/#pending-tests), and reported as such."

Comment (Qard) [reply packages/dd-trace/test/dogstatsd.spec.js:228]: Pending tests are ones without a test function, not ones marked skip. Those are two different concepts.

Comment (bengl) [reply packages/dd-trace/test/dogstatsd.spec.js:228]: Already `.only` still has a test running N times due to having N versions tested. You could argue that it's the same test, and therefore semantically correct, but I really think this is mostly just hair splitting.

To be honest: Pending isn't the worst thing here, since nothing should really stop us from supporting named pipes on Windows, but we haven't tested it anywhere else and the main reason I was skipping it is I ran into permission issues on GH Actions, and I decided it wasn't worth figuring out right now.

File: packages/dd-trace/src/dogstatsd.js (hunk 9 above)
Comment (bengl) [reply packages/dd-trace/src/dogstatsd.js:84]: I've cleaned this up quite a bit in the next push.

File: packages/dd-trace/src/metrics.js (hunk 11 above)
Comment (bengl) [reply packages/dd-trace/src/metrics.js:57]: It's not. The presence of `metricsProxyUrl` determines whether or not things will go to `request` at all.

**Review by Qard — APPROVED** [review]

File: packages/dd-trace/test/dogstatsd.spec.js (hunk 12 above)
Comment (rochdev) [reply packages/dd-trace/test/dogstatsd.spec.js:228]: This was really just a nitpick and I didn't expect it to become a larger conversation. Let's keep it as it is.

**Review by rochdev — APPROVED** [review]

**Review by Qard — APPROVED** [review]
